lifecycle 2.9.0-alpha08

terrakok opened the pull request:
Apply lifecycle 2.9.0-alpha08 from AOSP (commit `46295bc0b75a16f452e8e0090e8de41073d4dbb6`)

## Release Notes
N/A

terrakok commented:
Builds are fail because `lifecycle-runtime` library was published with Kotlin 2.1.0 for native targets with no KotlinTarget 1.9 compatibility. And we cannot build `lifecycle-runtime-compose` which depends on that.
```
.../lifecycle-runtime-linuxX64Main-2.9.0-alpha08.klib. 
Incompatible abi version. The current default is '1.8.0', found '1.201.0'. The library produced by 2.1.0 compiler
```

MatkovIvan commented on lifecycle/lifecycle-common/build.gradle:
```diff
@@ -103,11 +85,23 @@ kotlin {
 androidx {
     name = "Lifecycle-Common"
-    publish = Publish.SNAPSHOT_AND_RELEASE
+    type = LibraryType.PUBLISHED_LIBRARY
```
Not sure if our publishing will be happy with this change

terrakok replied:
as I see it doesn't affect our publishing

MatkovIvan reviewed (commented):
I see that build fails not only because kotlin version:
```
e: lifecycle/lifecycle-viewmodel-savedstate/src/commonMain/kotlin/androidx/lifecycle/SavedStateHandleSupport.kt:64:18 Unresolved reference: SavedState
e: lifecycle/lifecycle-viewmodel-savedstate/src/commonMain/kotlin/androidx/lifecycle/SavedStateHandleSupport.kt:102:23 Not enough information to infer type variable T
e: lifecycle/lifecycle-viewmodel-savedstate/src/commonMain/kotlin/androidx/lifecycle/SavedStateHandleSupport.kt:152:32 Unresolved reference: SavedState
```

terrakok commented:
ab

> I see that build fails not only because kotlin version:
> 
> ```
> e: lifecycle/lifecycle-viewmodel-savedstate/src/commonMain/kotlin/androidx/lifecycle/SavedStateHandleSupport.kt:64:18 Unresolved reference: SavedState
> e: lifecycle/lifecycle-viewmodel-savedstate/src/commonMain/kotlin/androidx/lifecycle/SavedStateHandleSupport.kt:102:23 Not enough information to infer type variable T
> e: lifecycle/lifecycle-viewmodel-savedstate/src/commonMain/kotlin/androidx/lifecycle/SavedStateHandleSupport.kt:152:32 Unresolved reference: SavedState
> ```

because it requires Savedstate update, yes. I will rebase it after

terrakok committed:
Replace build.gradle and src directories in lifecycle-common, lifecycle-runtime, lifecycle-viewmodel, lifecycle-viewmodel-savedstate with AOSP 2.9.0-alpha08 version.

AOSP commit: 46295bc0b75a16f452e8e0090e8de41073d4dbb6

terrakok committed:
Add web, wachos and tvos targets, use annotation-internal instead of androidx in lifecycle modules.

terrakok committed:
Replace invalid imports for proper KMP support

terrakok committed:
Add @JsName annotation to savedStateProvider method to fix compilation

terrakok committed:
Add web specific actuals for lifecycle modules.

terrakok committed:
Extract AbstractSavedStateViewModelFactory to common and deprecate methods with core.Bundle

terrakok committed:
Use `savedState` instead of `bundleOf` for state saving.

This change replaces `bundleOf` with `savedState` to save state in `SavedStateHandle`.

terrakok committed:
Dump API for lifecycle-viewmodel-savedstate

terrakok commented:
Well, the MR is ready and waits a solution to the klib abi version problem.

terrakok committed:
Temporary pin navigation-common dependencies until migrate on the new lifecycle and savedsate

terrakok commented:
I hope the CI will be green. If so, then Im ready to lend SavedState and Lifecycle modules to the jb-main branch

terrakok commented:
<img width="800" alt="image" src="https://github.com/user-attachments/assets/7f1d099a-89e1-4577-b498-4ace06ea0f10" />
the demo is being compiled and run

MatkovIvan commented on navigation/navigation-common/build.gradle:
```diff
@@ -55,12 +55,12 @@ androidXMultiplatform {
-                api project(":lifecycle:lifecycle-viewmodel")
-                api project(":lifecycle:lifecycle-viewmodel-savedstate")
+                api "org.jetbrains.androidx.core:core-bundle:1.1.0-alpha02"
+                api "org.jetbrains.androidx.core:core-uri:1.1.0-alpha02"
```
Why do we unpin it here?

igordmn replied:
We shouldn't pin to not stable versions

igordmn replied:
The current algorithm of unpinning is to find all project dependencies after a stable release, and replace them by the stable version.

If we pin to a not stable version, we miss dependencies

terrakok replied:
this is a temporary short term solution before a proper migration to new modules

igordmn replied:
Please, create a Major task with target Compose 1.8.0 to pin to the new stable versions. 

If we won't release new stable versions of these libs with Compose, then we'll change the target to 1.9

MatkovIvan reviewed (approved):
good to go into `integration`

igordmn commented:
Is there a commit that merges the state of `lifecycle 2.9.0-alpha08` (only resolving conflicts)? Otherwise it is hard to separate your changes and common main code.

terrakok commented:
> Is there a commit that merges the state of `lifecycle 2.9.0-alpha08` (only resolving conflicts)? Otherwise it is hard to separate your changes and common main code.

the first commit is a copy-paste from AOSP

igordmn commented:
> the first commit is a copy-paste from AOSP

I see. Was confused by the title, and at first thought it was not related 🙃
